gusto/transport_terms.py

```diff
@@ -0,0 +1,205 @@
+from firedrake import dx, dot, grad, div, jump, avg, inner, \
+    outer, sign, cross, CellNormal, curl, dS
+from gusto.terms import Term
+
+
+__all__ = ["LinearAdvectionTerm", "AdvectionTerm", "VectorInvariantTerm"]
+
+
+class TransportTerm(Term):
+    """
+    Base class for transport terms.
+
+    The equation is assumed to be in the form:
+
+    q_t + L(q) = 0
+
+    where q is the (scalar or vector) field to be solved for.
+
+    :arg state: :class:`.State` object.
+    :arg test: :class:`.TestFunction` object.
+    :arg ibp: string, stands for 'integrate by parts' and can take the value
+              None, "once" or "twice". Defaults to "once".
```

Comment:
Prefer an `enum`:

```python
from enum import Enum

class TransportTerm(Term):
   class IntegrateByParts(Enum):
      NO
      ONCE
      TWICE
   def __init__(self, ..., ibp):
      ...

...

transport = TransportTerm(..., TransportTerm.IntegrateByParts.ONCE)
```

Reply:
This way, you get error checking of values and extensibility somehow "for free".

Reply:
BTW, everyone points out to me that the zero-equivalent of ONCE is NEVER.